### ISSUE TITLE
Extend frontmatter links to support other actions

### ISSUE THREAD

**Issue by telenieko** [imdone/imdone#34]
I've just started using imdone (I'm on day 1 of the trial!) so maybe there are other ways to achieve what follows.

I keep pretty much everything in one folder tree, which means that for imdone it is all a single project, which is ok.

Now, how would one go with: "show me only cards from file XX".

That can currently be achieved by using a context for the file (setting it in the front matter). But a more powerful solution would be ...

Cards can have links at the top, those come from the frontmatter `links` section. Though those links currently only are that: links. But they could be much more:

  - They could allow to set a filter, just as clicking a tag sets a filter for the tag. Clicking such a link could modify the filter.
  - They could manipulate the card in magic ways (push due date x days, etc)

Crude example (missing the icons):

```yaml
links:
  - title: Focus on card's file
     filter: source.path = "${source.path}"
  - title: Push due date +7 days
     patch: due = due + "7 days"
  - title: Focus on cards of year ${meta.created.year}
     filter: source.path = /journal-${meta.created.year}/
```

Extra points if those links can be set in config.yml for the whole project.
Maybe those links would be on a `actions` section better than `links`.

**Describe alternatives you've considered**
- Setting up a global context for the file and filtering with it. Risk of duplicate contexts, and path does not always equal context.
- No alternative for the `patch` example.

**Comment by piascikj** [comment]
@telenieko I started working on this yesterday. What I came up looks something like this.
```yaml
    links:
      - href: >-
          https://twitter.com/intent/tweet?text=${encodedText}%0ATweeted%20with%20@imdoneio
        icon: fa-twitter
        pack: fab
        title: Tweet this card
      - action: 'filter source.path = "${source.path}"'
        icon: filter
        pack: fas
        title: 'Filter by ${source.path.split("/").pop()}'
      - action: 'newCard ${source.path}'
        icon: plus-square
        pack: fas
        title: 'New card in ${source.path.split("/").pop()}'
```

Haven't implemented patch though.  That may take some time.

Here's a couple of UI pics. Should be ready to release in a couple days.

<img width="406" alt="Screen Shot 2020-12-02 at 10 46 31 AM" src="https://user-images.githubusercontent.com/233505/100895872-41883780-347b-11eb-9595-3cc77a6f51b4.png">
<img width="405" alt="Screen Shot 2020-12-02 at 10 47 07 AM" src="https://user-images.githubusercontent.com/233505/100895906-4a790900-347b-11eb-81f0-43e836967184.png">

**Comment by telenieko** [comment]
This looks very well. thanks for the quick response and implementation!

**Comment by telenieko** [comment]
> Haven't implemented patch though. That may take some time.

On the `patch` thing, take a look also at #28 where I propose something similar when dragging cards into lists, probably some code can be shared between both features.

About that `newCard` on your example. I do not recall sein

**Comment by piascikj** [comment]
I just released 1.6.10 with some of the link functionality we've been discussing. Here's a sample of my **config.yml** to help you. It's currently an undocumented feature, but I'll be working on the docs while I shore it up and add other actions.
```yaml
keepEmptyPriority: false
code:
  include_lists:
    - TODO
    - DOING
    - DONE
    - PLANNING
    - FIXME
    - ARCHIVE
    - HACK
    - CHANGED
    - XXX
    - IDEA
    - NOTE
    - REVIEW
    - URGENT
    - TWEET
    - TODAY
    - READY
    - RELEASED
    - NEXT
    - NEW
    - NOW
    - CURRENT
    - ABANDONED
    - MORE
lists:
  - name: NOTE
    hidden: false
    ignore: false
  - name: What's Due?
    hidden: true
    ignore: false
    filter: >-
      dueDate < "${in 15 days}" AND list != READY AND list != RELEASED AND list
      != DONE +dueDate +order
    id: 0
  - name: TODO
    hidden: true
    ignore: false
  - name: NEXT
    hidden: true
    ignore: false
  - name: CURRENT
    hidden: false
    ignore: false
  - name: DOING
    hidden: false
    ignore: false
  - name: DONE
    hidden: false
    ignore: true
  - name: RELEASED
    hidden: true
    ignore: true
  - name: Recently Completed
    hidden: false
    ignore: false
    filter: 'completedDate > "${14 days ago}" -completed'
    id: 1
  - name: ABANDONED
    hidden: false
    ignore: true
settings:
  journalType: Folder
  journalPath: notes
  taskPrefix: '##'
  maxLines: 4
  addNewCardsToTop: true
  newCardSyntax: MARKDOWN
  cards:
    colors:
      - color: black
        filter: list=TODO
      - color: green
        filter: list=DOING OR tags=roadmap
      - color: red
        filter: allTags=bug AND list!=READY AND list!=DONE AND list!=RELEASED
      - color: black
        filter: placeholder
      - color: red
        filter: allTags=serial-codes
      - color: yellow
        filter: 'allTags=/1\.[6-9x]/'
      - color: red
        filter: tags = red
      - color: yellow
        filter: tags = yellow
      - color: green
        filter: tags = green
      - color: black
        filter: tags = black
      - color: red
        filter: allTags=urgent
    computed:
      date: (new Date()).toDateString()
      time: (new Date()).toLocaleTimeString()
      timestamp: (new Date()).toISOString()
      dueEmoji: !<tag:yaml.org,2002:js/function> |-
        function anonymous(
        ) {
        const due = this.totals["What's Due?"]
          let emoji = ':2nd_place_medal:'
          if (due >= 3) {
            emoji = ':fire:'
          } else if (due === 0) {
            emoji = ':rocket:'
          }
          return `<span style="font-size: 1.5em;">${emoji}</span>`
        }
      recentEmoji: !<tag:yaml.org,2002:js/function> |-
        function anonymous(
        ) {
        const recentlyCompleted = this.totals["Recently Completed"]
          let emoji = ':2nd_place_medal:'
          if (recentlyCompleted >= 3) {
            emoji = ':rocket:'
          } else if (recentlyCompleted === 0) {
            emoji = ':fire:'
          }
          return `<span style="font-size: 1.5em;">${emoji}</span>`
        }
      wipEmoji: !<tag:yaml.org,2002:js/function> |-
        function anonymous(
        ) {
        const doing = this.totals["DOING"]
          let emoji = ':2nd_place_medal:'
          if (doing >= 3) {
            emoji = ':fire:'
          } else if (doing === 0) {
            emoji = ':sleeping:'
          } else if (doing === 1) {
            emoji = ':rocket:'
          }
          return `<span style="font-size: 1.5em;">${emoji}</span>`
        }
    links:
      - href: >-
          https://twitter.com/intent/tweet?text=${encodedText}%0ATweeted%20with%20@imdoneio
        icon: fa-twitter
        pack: fab
        title: Tweet this card
      - action: !<tag:yaml.org,2002:js/function> |-
          function anonymous(
          ) {
            return this.actions.filter(`source.path = "${this.source.path}"`)
          }
        icon: filter
        pack: fas
        title: 'Filter by ${source.path.split("/").pop()}'
      - action: !<tag:yaml.org,2002:js/function> |-
          function anonymous(
          ) {
            return this.actions.newCard(this.source.path)
          }
        icon: plus-square
        pack: fas
        title: 'Add card in ${source.path.split("/").pop()}'
    props:
      hashtags: >-
        #100daysofcode #programmer #developer #hacking #markdown #Productivity
        #indiehackers #blogger
    template: |

      <!-- 
      expand:1
      created:${timestamp}
      -->
  doneList: DONE
  filterHistory:
    - allTags = "1.6.10"
    - allTags = "1.6.x"
    - source.path = "notes/2020-09/2020-09-28.md"
    - allTags = "urgent"
    - allTags = "customer"
    - allMeta.expand = "1"
  filteredLists:
    - filter: >-
        dueDate < "${in 15 days}" AND list != READY AND list != RELEASED AND
        list != DONE +dueDate +order
      hidden: false
      name: What's Due?
    - filter: 'completedDate > "${14 days ago}" -completed'
      name: Recently Completed
```

**Comment by telenieko** [comment]
The UI looks good.
Configuration is a bit messy (no user would be able to come up with the right syntax unless you have an example to follow).

This:
```js
          function anonymous(
          ) {
            return this.actions.filter(`source.path = "${this.source.path}"`)
          }
```

Could be:
```yaml
   filter: source.path = "${source.path}"
```

So, from the initial comment, the `filter` in config could be translated to `this.actions.filter` and make for a much cleaner/simpler usage aligned with `filteredLists` and other usages of filters.

The syntax for the `newCard` should also be available as a separate config element (`action`?). I think `this.actions` is undocumented, but maybe some of it can be documented and exposed like:

```yaml
  action: newCard(...)
```

Then, this.actions.filter is exposed as  `filter` because of it being common and to make it's usage similar to `filteredLists`, etc. Other actions are available in `action` ?

In the meantime, the functionality works, which is awesome!

**Comment by telenieko** [comment]
I see this has now been released in 1.6.23 with very nice and clean syntax.
Thank you!

**Comment by tf13** [comment]
Two related items: I can't get the custom links to display icons; and it would be nice to be able to define links across projects, rather than on a project-by-project basis. 

1. Icons don't display. Relevant section of config (with paths obscured):

```
links:
      - href: 'file:///path/to/directory/${text}'
        icon: folder
        pack: fas
        title: Open enclosing folder
      - href: >-
          file:///path/to/directory/${text}/todo.txt
        icon: list
        pack: fas
        title: Open task list
      - href: 'file://${source.repoId}'
        icon: skull
        pack: fas
        title: '${source.repoId}'
```

Here's the result — note that the hover text works; clicking on it also works — but the icon doesn't display:

<img width="304" alt="Screen Shot 2021-03-09 at 1 22 14 PM" src="https://user-images.githubusercontent.com/16142258/110518786-8ce55e00-80da-11eb-92ec-398ff4444d5d.png">

Not sure what I'm doing wrong, or if it's a bug.

2. It would be great to be able to set these links in config.json and have them appear in all projects.

**Comment by piascikj** [comment]
> Two related items: I can't get the custom links to display icons; and it would be nice to be able to define links across projects, rather than on a project-by-project basis.
> 
> 1. Icons don't display. Relevant section of config (with paths obscured):
> 
> ```
> links:
>       - href: 'file:///path/to/directory/${text}'
>         icon: folder
>         pack: fas
>         title: Open enclosing folder
>       - href: >-
>           file:///path/to/directory/${text}/todo.txt
>         icon: list
>         pack: fas
>         title: Open task list
>       - href: 'file://${source.repoId}'
>         icon: skull
>         pack: fas
>         title: '${source.repoId}'
> ```
> 
> Here's the result — note that the hover text works; clicking on it also works — but the icon doesn't display:
> 
> <img alt="Screen Shot 2021-03-09 at 1 22 14 PM" width="304" src="https://user-images.githubusercontent.com/16142258/110518786-8ce55e00-80da-11eb-92ec-398ff4444d5d.png">
> 
> Not sure what I'm doing wrong, or if it's a bug.
> 
> 1. It would be great to be able to set these links in config.json and have them appear in all projects.

@tf13 All font-awesome fas and fab icons will be available in 1.15.0.  Would you mind creating a separate feature request for (2)?
